snuba/query/processors/arrayjoin_optimizer.py

```diff
@@ -0,0 +1,300 @@
+from typing import Optional, Sequence, Set, cast
+
+from snuba.clickhouse.processors import QueryProcessor
+from snuba.clickhouse.query import Query
+from snuba.query.conditions import (
+    BooleanFunctions,
+    ConditionFunctions,
+    get_first_level_conditions,
+    in_condition,
+    is_binary_condition,
+    is_in_condition_pattern,
+)
+from snuba.query.dsl import arrayElement
+from snuba.query.expressions import Argument
+from snuba.query.expressions import Column as ColumnExpr
+from snuba.query.expressions import Expression
+from snuba.query.expressions import FunctionCall as FunctionCallExpr
+from snuba.query.expressions import Lambda
+from snuba.query.expressions import Literal as LiteralExpr
+from snuba.query.matchers import Any, Column, FunctionCall, Literal, Or, Param, String
+from snuba.request.request_settings import RequestSettings
+
+tags_key_pattern = FunctionCall(
+    None, String("arrayJoin"), (Column(None, None, String("tags.key")),)
```

Comment:
Do you mean the behavior of matching a column present in an arrayJoin function call ?

Reply:
It seems possible (maybe not likely, but possible) that any mapping-like `Nested` column (e.g. `contexts.*`) could have an `arrayJoin` applied to it which could be optimized in the same fashion. Or am I overlooking something, and this truly unique to the concept of tags and not generically any key/value mapping?

Reply:
This can in principle be applied to any key-value pair column. As a matter of fact I am still not sure why this was not applied to contexts to begin with (https://github.com/getsentry/snuba/blob/master/snuba/datasets/tags_column_processor.py#L111).
I think one of the reasons was that (fortunately) there is no such a thing as a `contexts_key` and `conntexts_value` keywords, still that did not prevent the user from actually send a valid `arrayJoin(contexts.key)` function, so that does not explain this enttirely.

I think the question is whether we want to encourage using `arrayJoin` at all.

For this PR I can make the column name parametric, though I would not go as far as wiring this processor for contexts as well unless we knew if there was a reason for not doing so.

Reply:
> I think the question is whether we want to encourage using `arrayJoin` at all.

I don't have a super strong opinion about `arrayJoin` in isolation. Personally, I'd rather encourage using `arrayJoin` directly if the only other option is `tags_{key,value}` since the behavior is at least a bit more obvious or explainable.

> For this PR I can make the column name parametric, 

I think this a good idea, looking at #983, I think the fact that there is an implicit schema requirement between the processor and the storage is problematic, regardless of whether or not other columns like `contexts` are included in the processor's responsibilities.

> though I would not go as far as wiring this processor for contexts as well unless we knew if there was a reason for not doing so.

I'm not sure how to parse this triple negative with much confidence I'm interpreting the intent correctly, can you rephrase this?

Reply:
"I'm not sure how to parse this triple negative with much confidence I'm interpreting the intent correctly, can you rephrase this?"

I would not apply this optimization to contexts just yet. None would be using it, and I am not 100% sure it would work just fine. I'd like to figure out why contexts were not supported i nthe first place.

Reply:
Ok, I suppose this makes sense, thanks for clarifying.

Reply:
made it parametric